filemanager/azureblobmanager.go

```diff
@@ -0,0 +1,368 @@
+package filemanager
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+
+	"github.com/rudderlabs/rudder-go-kit/logger"
+)
+
+type AzureBlobConfig struct {
+	Container      string
+	Prefix         string
+	AccountName    string
+	AccountKey     string
+	SASToken       string
+	EndPoint       *string
+	ForcePathStyle *bool
+	DisableSSL     *bool
+	UseSASTokens   bool
+}
+
+// NewAzureBlobManager creates a new file manager for Azure Blob Storage
+func NewAzureBlobManager(config map[string]interface{}, log logger.Logger, defaultTimeout func() time.Duration) (*azureBlobManager, error) {
+	return &azureBlobManager{
+		baseManager: &baseManager{
+			logger:         log,
+			defaultTimeout: defaultTimeout,
+		},
+		config: azureBlobConfig(config),
+	}, nil
+}
+
+func (manager *azureBlobManager) ListFilesWithPrefix(ctx context.Context, startAfter, prefix string, maxItems int64) ListSession {
+	return &azureBlobListSession{
+		baseListSession: &baseListSession{
+			ctx:        ctx,
+			startAfter: startAfter,
+			prefix:     prefix,
+			maxItems:   maxItems,
+		},
+		manager: manager,
+	}
+}
+
+// Upload passed in file to Azure Blob Storage
+func (manager *azureBlobManager) Upload(ctx context.Context, file *os.File, prefixes ...string) (UploadedFile, error) {
+	containerURL, err := manager.getContainerURL()
+	if err != nil {
+		return UploadedFile{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, manager.getTimeout())
+	defer cancel()
+
+	if manager.createContainer() {
+		_, err = containerURL.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
+		err = manager.suppressMinorErrors(err)
+		if err != nil {
+			return UploadedFile{}, err
+		}
+	}
+
+	fileName := path.Join(manager.config.Prefix, path.Join(prefixes...), path.Base(file.Name()))
+
+	// Here's how to upload a blob.
+	blobURL := containerURL.NewBlockBlobURL(fileName)
+	_, err = azblob.UploadFileToBlockBlob(ctx, file, blobURL, azblob.UploadToBlockBlobOptions{
+		BlockSize:   4 * 1024 * 1024,
+		Parallelism: 16,
```

Comment:
[optional] these look like good candidates for options although whoever might need to change them could simply do a PR :stuck_out_tongue: